### DISCUSSION TITLE
Suggestion: allow destructuring tuple in function parameters

### DISCUSSION THREAD

**Discussion by sporto** [gleam-lang/gleam#1451] [Ideas & suggestions]
I found myself doing this very often, which feels verbose

```
fn something(tuple) {
  let #(a, b) = tuple
  ...
}
```

Would be great to be able to do:

```
fn something(#(a, b)) {
   ...
}
```

And maybe if you want to keep the tuple as a parameter:

```
fn something(#(a, b) as tuple) {
   ...
}
```

Thanks

**Comment by lpil** [comment]
I'm not immediately in favour of this because it seems like another way to get values from tuples without much of an obvious advantage over the methods we have today. I'd like to be quite conservative with features in Gleam, only adding things if they enable something new or provide a big UX improvement.

For me this doesn't meet that mark, but perhaps it will resonate more with the community. I'll convert this to a discussion so people can share their thoughts.

RE that example, you can also do this

```rust
fn something(tuple: #(a, b)) -> a {
  tuple.0
}
```

**Comment by michallepicki** [comment]
From my point of view tuples (and one-constructor custom types) would be the only types where this can be done in the function head, without pushing for multiple function clauses support. I would find this mostly useful in initial stages of code development, where I'm prototyping some API and want to "just" pass some data around, and expect the fields to change. So I think it shouldn't cause technically any issues and I do agree with @sporto that it makes code a tiny bit more concise in that particular scenario.

But it would be "different" than any other type...

Alternatives that require only a bit more typing are:
- using `.0`, `.1` etc (disadvantage: fields aren't named for readability)
- passing multiple arguments separately instead (the disadvantage here is that it doesn't work nicely with the pipeline `|>` operator)
- destructuring the tuple at the beginning of the function
- defining your own one-constructor custom type

I personally think destructuring at the beginning of the function is not that much of a hassle, allows to give names to the fields, and has a very natural and easy code refactoring transition to a custom type.

**Comment by bfeitknecht** [comment]
How about some inspiration from haskell's as-pattern syntax?
```gleam
fn f(t@(a, b): #(A, B)) -> Nil {
  echo t
  echo a
  echo b
}
```

**Reply by lpil** [reply]
We already have that pattern, the `as` keyword!

**Comment by inoas** [comment]
I am not in favour of encouraging to use tuples (I consider them unnamed singleton types) over gleam records.

But even for records, I think keeping the function signature easy to read is worth a lot and if type signatures are added things are complex enough to grasp.

**Comment by JosephTLyons** [comment]
More generically, disallowing function parameters to be destructured in place seems a bit inconsistent when looking at what is allowed in `use` syntax.

Consider the following function:

```gleam
fn to_bool(int: Int, cast: fn(Int) -> Bool) -> Bool {
  cast(int)
}
```

Then these two forms are both valid:

```gleam
pub fn main() {
  to_bool(1, fn(int) {
    int > 0
  })
}
```

```gleam
pub fn main() {
  use int <- to_bool(1)
  int > 0
}
```

However, given this new type and function:

```gleam
type NewInt {
  NewInt(Int)
}

fn to_bool(int: NewInt, cast: fn(NewInt) -> Bool) -> Bool {
  cast(int)
}
```

We are able to destructure in `use`:

```gleam
pub fn main() {
  let int = NewInt(1)
  use NewInt(int) <- to_bool(int)
  int > 0
}
```

But not in the non-`use` form:

```gleam
pub fn main() {
  let int = NewInt(1)
  to_bool(int, fn(NewInt(int)) { // Syntax error
    int > 0
  })
}
```

Which means you don't get the nice `convert {to,from} use` code action when you have it in the valid form using `use`.

I would think that if `use` and non`-use` are supposed to be equivalent forms, Gleam should either disallow destructuring in `use` or allow it in function args; I tend to prefer the later. 😅

**Reply by lpil** [reply]
Use isn't a one to one mapping onto the function call and anonymous function syntax, it also includes pattern matching.

The use syntax and the current limitations of language server code actions don't imply any additional features are needed for other language constructs.

**Comment by vistuleB** [comment]
I often found myself wishing for this feature while writing lambda functions for the case of tuple destructuring but I think there's a pretty strong argument it will make code less legible overall, even if shorter by 1 line:

- for single-constructor types, allowing destructing inside the function argument would introduce a brand new type of situation in which an uppercase token can appear; all the information a reader currently gets a glance by looking at what is uppercase and what is not will be polluted/modulated by this new pattern
- for tuples, the number of arguments to the function would no longer be visible at a glance; the reader's current task when reading arguments is a 1-to-1 matching task (what is this one? what is this one? etc) but if we allowed destructuring tuples the reader's task would become grouping-and-then-matching, i.e. a parsing step is added, I'm not sure it's a win

**Reply by lpil** [reply]
In Gleam we don't have multiple ways to do the same thing, and we only add features to solve problems that cannot otherwise be solved, so this feature is unsuitable for addition.

**Reply by vistuleB** [reply]
If this is your final decision then doesn't it close this discussion? Shouldn't you be replying to one of the original two "pro" comments above?

(I'm arguing on your side against the suggestion)

**Reply by lpil** [reply]
Yup, there's not much reason to continue to discuss this proposal.

**Comment by inoas** [comment]
I love that the function signatures are easy to spot. I sometimes wish I could drop the indentation around single `function => case` wrappings (which would then be closer to having multiple function heads) but I am glad we do not have destructing in function heads. Yes it means a little bit more boilerplate but it also means I can focus on all the primary things a function offers:
function name, input types, output types, labels.

**Comment by kyleshevlin** [comment]
I'm brand new to the language and I was honestly surprised I couldn't destructure the tuple in that way. Before I say the following, I want to be clear, I'm not suggesting you should acquiesce to me or anyone. I have no clue what would be best for the language. Just want to share my thought.

Basically, I agree with those who want to destructure early for clarity/legibility. I'm coming from a JavaScript/TypeScript background. We don't have tuples, but we can destructure arrays. And with TypeScript's `const` keyword, you can get tuple types and therefore do this destructuring with type safety.

```typescript
function makeGreeting([name, location]: [string, string]) {
  // ...
}
```

I expected to be able to do the same in Gleam.

In JS/TS, I'm not a fan of "just remember what is in what position". Reading: `tuple[0]` vs `tuple[1]` doesn't really jog my memory or do anything for providing context. I feel the same way about Gleam and using `tuple.0` or `tuple.1`.

I'll probably destructure what I need in the fn body for now. Just how I like to work, for whatever that might be worth.

Thanks for lang. It's really cool, been having fun learning it.
